extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,618 @@
+                             !!sfConfig['expandByOverlay']) || true;
+
+    /** @private {boolean} */
+    this.expandByPush_ = (sfConfig.hasOwnProperty('expandByPush') &&
+                          !!sfConfig['expandByPush']) || true;
+
+    this.registerSafeframeHost();
+  }
+
+  /**
+   * Returns the Safeframe specific name attributes that are needed for the
+   * Safeframe creative to properly setup.
+   * @return {!JsonObject}
+   */
+  getSafeframeNameAttr() {
+    const attributes = dict({});
+    attributes['uid'] = this.uid_;
+    attributes['hostPeerName'] = this.win_.location.origin;
+    attributes['initialGeometry'] = this.getInitialGeometry();
+    attributes['permissions'] = JSON.stringify(
+        dict({
+          'expandByOverlay': this.expandByOverlay_,
+          'expandByPush': this.expandByPush_,
+          'readCookie': false,
+          'writeCookie': false,
+        }));
+    attributes['metadata'] = JSON.stringify(
+        dict({
+          'shared': {
+            'sf_ver': this.baseInstance_.safeframeVersion,
+            'ck_on': 1,
+            'flash_ver': '26.0.0',
+          },
+        }));
+    attributes['reportCreativeGeometry'] = this.isFluid_;
+    attributes['isDifferentSourceWindow'] = false;
+    attributes['sentinel'] = this.sentinel_;
+    return attributes;
+  }
+
+  /**
+   * Returns the initialGeometry to assign to the name of the safeframe
+   * for rendering. This needs to be done differently than all the other
+   * geometry updates, because we don't actually have access to the
+   * rendered safeframe yet. Note that we are using getPageLayoutBox,
+   * which is not guaranteed to be perfectly accurate as it is from
+   * the last measure of the element. This is fine for our use case
+   * here, as even if the position is slightly off, we'll send the right
+   * size.
+   * @return {string}
+   */
+  getInitialGeometry() {
+    const ampAdBox = this.baseInstance_.getPageLayoutBox();
+    const heightOffset = (ampAdBox.height - this.creativeSize_.height) / 2;
+    const widthOffset = (ampAdBox.width - this.creativeSize_.width) / 2;
+    const iframeBox = {
+      top: ampAdBox.top + heightOffset,
+      bottom: ampAdBox.bottom - heightOffset,
+      left: ampAdBox.left + widthOffset,
+      right: ampAdBox.right - widthOffset,
+      height: this.creativeSize_.height,
+      width: this.creativeSize_.width,
+    };
+    return this.formatGeom_(iframeBox);
+  }
+
+  /**
+   * Registers this instance as the host API for the current sentinel.
+   * If the global safeframe listener has not yet been created, it creates
+   * that as well.
+   */
+  registerSafeframeHost() {
+    dev().assert(this.sentinel_);
+    safeframeHosts[this.sentinel_] = safeframeHosts[this.sentinel_] || this;
+    if (!safeframeListenerCreated_) {
+      safeframeListenerCreated_ = true;
+      this.win_.addEventListener('message', safeframeListener, false);
```

Comment:
You need to remove this listener when unlayoutCallback is called on the parent amp-ad element.  Create a cleanup function that you call which removes listener.

Reply:
done